### PR TITLE
Fix and refactor AMT

### LINKED ISSUE #4775: Fix erroneous federal AMT calculation
The following simple 2022 test case illustrates a major flaw in the federal AMT calculations that can cause PolicyEngine-US to generate federal income tax amounts that are many thousands of dollars larger than the correct amounts.  This flaw causes errors as far back at 2018 (see #4769).

The characterization of this situation as a bug seems sensible because online [TAXSIM-35](https://taxsim.nber.org/taxsim35) and [Tax-Calculator 4.2.0](https://github.com/PSLmodels/Tax-Calculator) and 2022 [Turbo Tax](https://turbotax.intuit.com) are all in agreement that this test case has a zero AMT.   PolicyEngine-US 1.28.0, however, generates a federal AMT amount equal to $7,538.50.

That this bug exists is not particularly surprising because the developers of the `alternative_minimum_tax` formula did not specify any unit tests for it.   (At least, I can't find any tests when doing a recursive grep operation over the source code tree.).  Also, the 2022 federal income testing done in issue #1773 never considered test cases with more than $10K in qualified dividends or more than $10K in long-term capital gains.

First, I show the YAML file containing the test case.  Then I show several screenshots from Turbo Tax processing a make-believe person like the test case.

**The YAML test case:**
```
- name: Single Texan with no children, $150K wages, $50K QDIV, $50K LTCG
  absolute_error_margin: 0.01
  period: 2022
  input:
    people:
      person1:
        age: 50
        employment_income: 150_000
        qualified_dividend_income: 50_000
        long_term_capital_gains: 50_000
    spm_units:
      spm_unit:
        members: [person1]
    tax_units:
      tax_unit:
        members: [person1]
    households:
      household:
        members: [person1]
        state_code: TX
  output:
    # expected output from online TAXSIM-35
    # which is at: https://taxsim.nber.org/taxsim35/
    #
    adjusted_gross_income: 250_000
    taxable_income: 237_050
    net_investment_income_tax: 1_900
    income_tax: 43_627.50

##############################################################################
# ABOVE TEST IS IN A FILE NAMED case22.yaml
# TEST ENVIRONMENT:
#   % pip list | grep policyengine
#   policyengine-core             2.21.8
#   policyengine-us               1.28.0
# RESULTS FROM RUNNING THIS TEST ARE AS FOLLOWS:
#
#=============================== FAILURES ===================================
#_____________________________ test session _________________________________
#/Users/mrh/work/.../case22.yaml:
#  Test 'Single Texan with no children, $150K wages, $50K QDIV, $50K LTCG':
#    income_tax@2022: [51166.] differs from 43627.5
#                     with an absolute margin [7538.5] > 0.01
#======================= short test summary info ============================
#FAILED case22.yaml::
#========================== 1 failed in 3.19s ===============================
#Trace output is in file: case22.trace
#% grep alternative_minimum_tax case22.trace
#         alternative_minimum_tax<2022, (default)> = [7538.5]
# So, the income tax difference is caused solely by the AMT difference.
```

**2022 Turbo Tax screenshots:**

Screenshot 2024-07-23 at 11 38 10 AM

.

Screenshot 2024-07-23 at 11 46 40 AM

.

And finally, federal Form 6251:

Screenshot 2024-07-23 at 11 52 24 AM

Screenshot 2024-07-23 at 11 52 45 AM

Screenshot 2024-07-23 at 11 53 23 AM

Screenshot 2024-07-23 at 11 54 18 AM

### PR DESCRIPTION
**Pull request by PavelMakarchuk**
Fixes #4775

### CODE REVIEW

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        regular_tax_before_credits = tax_unit(
+            "regular_tax_before_credits", period
+        )
+        lump_sum_distributions = tax_unit(
+            "form_4972_lumpsum_distributions", period
+        )
+        tax_unit("capital_gains_tax", period)
+        tax_before_credits = (
+            regular_tax_before_credits + regular_tax_before_credits
+        )
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py:125]: This is the piece that was added in this PR

Comment (martinholmer) [reply policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py:125]: Can you explain why `regular_tax_before_credits` is being added to itself?
What am I missing?

**Comment by martinholmer** [comment]
@PavelMakarchuk,  in PR #4815 the new code looks like this:

Screenshot 2024-07-31 at 10 34 52 AM

I don't understand line 122: the capital_gains_tax is not being assigned to a local variable.
What's the purpose of line 122?

And on line 124, I don't understand the point of adding `regular_tax_before_credits`  to itself?

Should the new code be something like this?
```
        tax_before_credits = (
            regular_tax_before_credits + tax_unit("capital_gains_tax", period)
        )
```

**Comment by PavelMakarchuk** [comment]
> @PavelMakarchuk, in PR #4815 the new code looks like this:
> 
> Screenshot 2024-07-31 at 10 34 52 AM
> I don't understand line 122: the capital_gains_tax is not being assigned to a local variable. What's the purpose of line 122?
> 
> And on line 124, I don't understand the point of adding `regular_tax_before_credits` to itself?
> 
> Should the new code be something like this?
> 
> ```
>         tax_before_credits = (
>             regular_tax_before_credits + tax_unit("capital_gains_tax", period)
>         )
> ```

Yes, my mistake, that was exactly the intention

**Comment by martinholmer** [comment]
@PavelMakarchuk,  There seems to be a problem with building the documentation.
What's that all about?

**Comment by PavelMakarchuk** [comment]
> @PavelMakarchuk, There seems to be a problem with building the documentation. What's that all about?

Yes the CI is affecting every Pull Request right now - we have an [issue](https://github.com/PolicyEngine/policyengine-us/issues/4761) filed for it but it has not been fixed yet

File: changelog_entry.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- bump: patch
+  changes:
+    fixed:
+    - Refactor the alternative minimum tax files.
+    - Inculde capital gains tax in the final alternative minimum tax calculation.
```
Comment (MaxGhenis) [inline changelog_entry.yaml:5]: indent other too
```suggestion
      - Include capital gains tax in the final alternative minimum tax calculation.
```

File: policyengine_us/tests/policy/baseline/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      tax_unit:
+        members: [person1]
+    households:
+      household:
+        members: [person1]
+        state_code: TX
+  output:
+    # expected output from online TAXSIM-35
+    # which is at: https://taxsim.nber.org/taxsim35/
+    #
```
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.yaml:24]:
```suggestion
```

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        line54 = min_(line48, line53)
+        cgtax2 = line54 * cg.rates["2"]
+        line56 = line47 + line54
+        line57 = where(line41 == line56, 0, line46 - line56)
+        linex2 = where(line41 == line56, 0, max_(0, line54 - line48))
+        cgtax3 = line57 * cg.rates["3"]
+        line61 = where(
+            line38 == 0,
+            0,
+            0.25
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py:106]: could you parameterize this while you're here?

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    entity = TaxUnit
+    definition_period = YEAR
+    label = "Alternative Minimum Tax"
+    unit = USD
+    documentation = "Alternative Minimum Tax (AMT) liability"
+
+    def formula(tax_unit, period, parameters):
+        amt_income = tax_unit("amt_income", period)
+        # Form 6251, Part II top
+        amt = parameters(period).gov.irs.income.amt
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py:15]: and elsewhere
```suggestion
        p = parameters(period).gov.irs.income.amt
```

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/amt_income.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class amt_income(Variable):
+    value_type = float
+    entity = TaxUnit
+    definition_period = YEAR
+    label = "AMT taxable income"
+    unit = USD
+    reference = "https://www.law.cornell.edu/uscode/text/26/55#b_2"
+
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/amt_income.py:11]: Could you break these out to simplify this as following:
```suggestion
    adds = ["taxable_income", "amt_excluded_deductions", "amt_separate_addition"]
```

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/regular_tax_before_credits.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        dwks36 = max_(0, dwks34 - dwks1)
+        dwks37 = max_(0, dwks33 - dwks36)
+
+        p = parameters(period).gov.irs.income
+
+        dwks38 = p.amt.capital_gains.capital_gain_excess_tax_rate * dwks37
+        # Break in worksheet lines
+        dwks39 = dwks19 + dwks20 + dwks28 + dwks31 + dwks37
+        dwks40 = dwks1 - dwks39
+        dwks41 = 0.28 * dwks40
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/regular_tax_before_credits.py:52]: please parameterize

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/regular_tax_before_credits.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            b = str(i)
+            bracket_top = bracket_tops[b][filing_status]
+            reg_tax += bracket_rates[b] * amount_between(
+                reg_taxinc, bracket_bottom, bracket_top
+            )
+            bracket_bottom = bracket_top
+
+        # Return to worksheet lines
+        dwks42 = reg_tax
+        dwks43 = sum(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/regular_tax_before_credits.py:70]: is `sum` vectorized? i think we might just need `+`

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: changelog_entry.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+- bump: patch
+  changes:
+    fixed:
+      - Refactor the alternative minimum tax files.
+      - Inculde capital gains tax in the final alternative minimum tax calculation.
```
Comment (MaxGhenis) [inline changelog_entry.yaml:5]:
```suggestion
      - Include capital gains tax in the final alternative minimum tax calculation.
```

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/income/amt_separate_addition.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    definition_period = YEAR
+    label = "AMT taxable income separate addition"
+    unit = USD
+    reference = "https://www.law.cornell.edu/uscode/text/26/55#b_2"
+
+    def formula(tax_unit, period, parameters):
+        taxable_income = tax_unit("taxable_income", period)
+        excluded_deductions = tax_unit("amt_excluded_deductions", period)
+        amt_inc = taxable_income + excluded_deductions
+        amt = parameters(period).gov.irs.income.amt
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/income/amt_separate_addition.py:16]: and elsewhere
```suggestion
        p = parameters(period).gov.irs.income.amt.exemption
```

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/income/amt_separate_addition.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+        reduced_amt_inc = max_(0, amt_inc - amt.exemption.separate_limit)
+        return (
+            max_(
+                0,
+                min_(
+                    amt.exemption.amount[filing_status],
+                    amt.exemption.phase_out.rate * reduced_amt_inc,
+                ),
+            )
+            * separate
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/income/amt_separate_addition.py:28]: just as a note for the future, we could speed this up after https://github.com/PolicyEngine/policyengine-us/issues/4846

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/income/amt_excluded_deductions.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    reference = "https://www.law.cornell.edu/uscode/text/26/55#b_2"
+
+    def formula(tax_unit, period, parameters):
+        itemizing = tax_unit("tax_unit_itemizes", period)
+        standard_deduction = tax_unit("standard_deduction", period)
+        salt_deduction = tax_unit("salt_deduction", period)
+        return where(
+            itemizing,
+            salt_deduction,
+            standard_deduction,
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/income/amt_excluded_deductions.py:19]: this should fit on one line

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

**Comment by MaxGhenis** [comment]
Seems like we're hitting some sort of resource limit on the GH Actions servers, around the microsim tests. I don't think this PR is adding significant resource use, but it could be a tipping point.

Some options I see include:
1. Trim down the microsim tests
2. Improve efficiency, e.g. using more `defined_for`s or caching
3. Break up the test suite (not sure how resource constraints are handled / if this would help)
4. Pay for [larger runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners)

@nikhilwoodruff any other options you'd see?

File: policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py (hunk 1 above)
Comment (MaxGhenis) [reply policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py]: is this resolved?

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/irs/tax/federal_income/alternative_minimum_tax/alternative_minimum_tax.py]: Yes

**Comment by PavelMakarchuk** [comment]
> Seems like we're hitting some sort of resource limit on the GH Actions servers, around the microsim tests. I don't think this PR is adding significant resource use, but it could be a tipping point.
> 
> Some options I see include:
> 
> 1. Trim down the microsim tests
> 2. Improve efficiency, e.g. using more `defined_for`s or caching
> 3. Break up the test suite (not sure how resource constraints are handled / if this would help)
> 4. Pay for [larger runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners)
> 
> @nikhilwoodruff any other options you'd see?

I can remove Microsim tests (mostly not entirely) if we agree that this is the solution that we are going for

**Comment by MaxGhenis** [comment]
Sure, try removing one or two to start?

Does `make test` work for you locally?

**Comment by PavelMakarchuk** [comment]
> > Seems like we're hitting some sort of resource limit on the GH Actions servers, around the microsim tests. I don't think this PR is adding significant resource use, but it could be a tipping point.
> > Some options I see include:
> > 
> > 1. Trim down the microsim tests
> > 2. Improve efficiency, e.g. using more `defined_for`s or caching
> > 3. Break up the test suite (not sure how resource constraints are handled / if this would help)
> > 4. Pay for [larger runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners)
> > 
> > @nikhilwoodruff any other options you'd see?
> 
> I can remove Microsim tests (mostly not entirely) if we agree that this is the solution that we are going for

I would also think the reducing the test suite for structural reforms could help

**Comment by PavelMakarchuk** [comment]
> Sure, try removing one or two to start?
> 
> Does `make test` work for you locally?

The `make test` passes locally

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**